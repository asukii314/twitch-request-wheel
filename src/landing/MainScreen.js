import {Component} from 'react';
import {Button, Modal} from 'react-bootstrap';
import ChatActivity, { ActivityStatus } from '../ChatActivity';
import ConfettiExplosion from '@reonomy/react-confetti-explosion';
import GameRequest from '../components/GameRequest'
import MessageHandler from '../MessageHandler';
import PlayerSelect from '../components/PlayerSelect';
import Sidebar from './Sidebar'
import WheelComponent from '../WheelComponent'; //'react-wheel-of-prizes'
import * as fakeStates from '../example-states';

import './MainScreen.css';
// import 'bootstrap/dist/css/bootstrap.css';
const randomColor = require('randomcolor');


export default class MainScreen extends Component {
    constructor(props){
        super(props);
        this.chatActivity = new ChatActivity(this.props.channel)
        this.state = {
            allowGameRequests: true,
            gameSelected: null,
            messages: {},
            colors: randomColor({count: 99, luminosity: 'light', hue: 'blue'}),
            counter: 0,
            history: [],
            nextGameIdx: 0,
            showPlayerSelect: false,
            showOptionsModal: false
        };

        this.playerSelector = null;
        this.messageHandler = null;

        this.changeNextGameIdx = this.changeNextGameIdx.bind(this);
        this.moveNextGameFwd = this.moveNextGameFwd.bind(this);
        this.moveNextGameBack = this.moveNextGameBack.bind(this);
        this.addGameRequest = this.addGameRequest.bind(this);
        this.toggleLock = this.toggleLock.bind(this);
        this.setNextGame = this.setNextGame.bind(this);
        this.addGameToQueue = this.addGameToQueue.bind(this);
        this.onWheelSpun = this.onWheelSpun.bind(this);
        this.removeGame = this.removeGame.bind(this);
        this.onMessage = this.onMessage.bind(this);
        this.toggleAllowGameRequests = this.toggleAllowGameRequests.bind(this);
        this.togglePlayerSelect = this.togglePlayerSelect.bind(this);
        this.routePlayRequest = this.routePlayRequest.bind(this);
        this.routeLeaveRequest = this.routeLeaveRequest.bind(this);
        this.routeOpenQueueRequest = this.routeOpenQueueRequest.bind(this);
        this.routeCloseQueueRequest = this.routeCloseQueueRequest.bind(this);
        this.routeClearQueueRequest = this.routeClearQueueRequest.bind(this);
        this.startGame = this.startGame.bind(this);
        this.setMessageHandlerRef = this.setMessageHandlerRef.bind(this);
        this.setPlayerSelectRef = this.setPlayerSelectRef.bind(this);
    }

    componentDidMount() {
        if (window.location.hash.indexOf('fakestate=true') !== -1) {
            if (window.location.hash.indexOf('playerselect=true') !== -1) {
                this.setState(
                    Object.assign({}, fakeStates.MainScreen, {
                        showPlayerSelect: true
                    })
                );
            } else {
                this.setState(fakeStates.MainScreen);
            }
        }
    }

    changeNextGameIdx = (delta = 1) => {
        if (this.state.nextGameIdx + delta > this.state.history.length) return false;
        if (this.state.nextGameIdx + delta < 0) return false;
        this.setState((state) => {
            return {
                nextGameIdx: state.nextGameIdx + delta
            }
        });
        return true;
    }

    changeGameOrder = (history, nextGameIdx) => {
        if (nextGameIdx > history.length) return false;
        this.setState({
            history,
            nextGameIdx
        });
        return true;
    }

    moveNextGameFwd = () => {
        return this.changeNextGameIdx();
    }

    moveNextGameBack = () => {
        return this.changeNextGameIdx(-1);
    }

    addGameRequest = (gameObj, user) => {
        this.setState((state) => {
            return {
                ...state,
                messages: {
                    ...this.state.messages,
                    [gameObj.longName]: {
                        ...gameObj,
                        username: user,
                        time: Date.now(),
                        locked: false,
                        chosen: false
                    }
                },
                counter: this.state.counter + 1
            };
        })
    }

    toggleLock = (game) => {
        const stateCopy = {...this.state.messages[game]};
        stateCopy.locked = !stateCopy.locked

        this.setState(() => {
            return {
                ...this.state,
                messages: {
                    ...this.state.messages,
                    [game]: stateCopy
                }
            }
        });
    }

    // @return: the number of games ahead of this one, after successfully inserting in queue
    // (i.e. if it's the very next game, return 0; if there's one ahead, return 1; etc)
    setNextGame = (gameObj) => {
        let idx = this.state.nextGameIdx;

        // insert next game at next up position by default, but
        //      *after* any other manually inserted games
        while (idx < this.state.history.length && this.state.history[idx]?.override) {
            idx++;
        }

        this.setState((state) => {
            return {
                ...state,
                history: [
                    ...state.history.slice(0, Math.max(0, idx)),
                    {
                        ...gameObj,
                        override: true,
                        time: Date.now()
                    },
                    ...state.history.slice(idx)
                ]
            }
        });

        return idx - this.state.nextGameIdx;
    }

    addGameToQueue = (gameObj) => {
        // update history + game card highlight color
        this.setState((state) => {
            return {
                ...state,
                gameSelected: gameObj,
                history: [
                    ...this.state.history,
                    {
                        ...gameObj,
                        override: false
                    }
                ],
                messages: {
                    ...state.messages,
                    [gameObj.longName]: {
                        ...state.messages[gameObj.longName],
                        chosen: true
                    }
                }
            }
        });
    }

    clearModal = () => {
        this.setState({
            gameSelected: null
        });
    }

    getGamesList = () => {
        return {
            allowedGames: this.messageHandler?.state.allowedGames,
            validGames: this.messageHandler?.state.validGames
        }
    }

    onWheelSpun = (gameLongName) => {
        const gameRequestObj = this.state.messages?.[gameLongName];
        if (!gameRequestObj) return;

        // send confirmation message in chat
        const requester = gameRequestObj.username;

        this.addGameToQueue(gameRequestObj);

        // remove card unless it's locked
        if (!this.state.messages[gameLongName].locked) {
            setTimeout(() => {
                this.removeGame(gameLongName);
                this.clearModal();
            }, 4000);
        }

        return this.chatActivity.getStatusPromise(requester).then((status) => {
            let msg = "";
            switch(status) {
                case ActivityStatus.DISCONNECTED:
                    msg = `/me ${gameRequestObj.name} just won the spin, but it doesn't seem like @${requester} is still around. Hope someone else wants to play!`
                    break;

                case ActivityStatus.ACTIVE:
                    msg = `/me @${requester}, good news - ${gameRequestObj.name} just won the spin!`;
                    break;

                case ActivityStatus.IDLE:
                default:
                    msg = `/me @${requester}, good news - ${gameRequestObj.name} just won the spin! (I hope you're still around!)`;
            }
            return this.messageHandler.sendMessage(msg);
        })

    }

    removeGame = (gameLongName) => {
        const newMessageObj = {...this.state.messages};
        delete newMessageObj[gameLongName];
        this.setState((state) => {
            return {
                ...state,
                messages: newMessageObj,
                counter: this.state.counter + 1
            };
        })
    }

    onMessage = (message, user, metadata) => {
        this.chatActivity.updateLastMessageTime(user);
    }

    toggleAllowGameRequests = (allow=null) => {
        let {allowGameRequests} = this.state;
        if (allow !== null) {
            allowGameRequests = !allow;
        }
        this.setState((state) => {
            return {
                allowGameRequests: !allowGameRequests
            }
        })
    }

    toggleOptionsModal = () => {
        this.setState((state) => {
            return {
                showOptionsModal: !state.showOptionsModal
            }
        })
    }

    togglePlayerSelect = () => {
        this.setState((state) => {
            return {
                showPlayerSelect: !state.showPlayerSelect
            }
        })
    }

    routePlayRequest = (user, {sendConfirmationMsg = true, isPrioritySeat = false}) => {
        const msg = this.state.showPlayerSelect
            ? this.playerSelector?.handleNewPlayerRequest(user, {isPrioritySeat})
            : 'sign-ups are currently closed; try again after this game wraps up!'

        if (sendConfirmationMsg) {
            this.messageHandler?.sendMessage(`/me @${user}, ${msg}`);
        }
    }

    routeLeaveRequest = (user) => {
        this.playerSelector?.removeUser(user);
    }

    routeOpenQueueRequest = () => {
        this.setState((state) => {
            return {
                ...state,
                showPlayerSelect: true
            }
        })
        this.playerSelector?.openQueue();
    }

    routeCloseQueueRequest = () => {
        this.playerSelector?.closeQueue();
    }

    routeClearQueueRequest = () => {
        this.playerSelector?.clearQueue();
    }

    startGame = () => {
        if (this.state.showPlayerSelect) {
            this.togglePlayerSelect();
            this.moveNextGameFwd();
            return true;
        }
        return false;
    }

    setMessageHandlerRef = (ps) => {
        this.messageHandler = ps;
    };

    setPlayerSelectRef = (mh) => {
        this.playerSelector = mh;
    };

    renderGameChosenModal(gameObj) {
        let confettiProps = {
            force: 0.6,
            duration: 3500,
            particleCount: 100,
            floorHeight: Math.max(window.outerWidth, window.outerHeight),
            floorWidth: Math.max(window.outerWidth, window.outerHeight)
        };
        let requestedBy;
        if (gameObj.username) {
            requestedBy = (<h4>requested by @{gameObj.username}</h4>);
        }
        return (
            <>
                <div className="overlay fade-in-out">
                    <div className="confetti-wrapper">
                        <ConfettiExplosion {...confettiProps} />
                    </div>
                </div>
                <div className="confetti-modal modal-game-chosen fade-in-out" onClick={()=>this.removeGame(gameObj.longName)}>
                    <h1>{gameObj.name}</h1>
                    {requestedBy}
                </div>
            </>
        );
    }

    renderOptionsModal = () => {
        let {allowedGames, validGames} = this.messageHandler.state;
        let gamePackList = [].concat(...Object.entries(validGames).map((packData, idx) => {
            return Object.keys(packData[1]).map(gameData => {
                let gameId = `${packData[0]} ${gameData}`.replace(/\W/ig, '_');
                return {
                    id: gameId,
                    game: gameData,
                    pack: packData[0]
                }
            })
        }))
        // let gamesList = gamePackList.map(g => g.game);
        console.log('gamePackList:', gamePackList, allowedGames);

        // return (
        //     <>
        //         <div className="overlay fade-in" onClick={this.toggleOptionsModal}></div>
        //         <div className="modal modal-options fade-in">
        //             <h2>Options</h2>
        //             <div className="options-list">
        //                 <ul>
        //                     {gamePackList.map(({id, game, pack}, idx) => {
        //                         // let gameId = `${g.pack} ${g.game}`.replace(/\W/ig, '_');
        //                         return (
        //                             <li key={id}>
        //                                 <input type="checkbox" id={id} name={id} value={id} /> <label htmlFor={id}>{pack}: {game}</label>
        //                             </li>
        //                         )}
        //                     )}
        //                 </ul>
        //             </div>
        //         </div>
        //     </>
        // );
        return (

            <Modal
                show={this.state.showOptionsModal}
                onHide={this.toggleOptionsModal}
                size="lg"
                aria-labelledby="contained-modal-title-vcenter"
                centered>
                <Modal.Header closeButton>
                    <Modal.Title id="contained-modal-title-vcenter">
                        Modal heading
                    </Modal.Title>
                </Modal.Header>
                <Modal.Body>
                    <h4>Options</h4>
                    <div className="options-list">
                        <ul>
                            {gamePackList.map(({id, game, pack}, idx) => {
                                // let gameId = `${g.pack} ${g.game}`.replace(/\W/ig, '_');
                                return (
                                    <li key={id}>
                                        <input type="checkbox" id={id} name={id} value={id} /> <label htmlFor={id}>{pack}: {game}</label>
                                    </li>
                                )}
                            )}
                        </ul>
                    </div>
                </Modal.Body>
                <Modal.Footer>
                    <Button onClick={this.toggleOptionsModal}>Close</Button>
                </Modal.Footer>
            </Modal>

        );
    }

    render() {
        const gameRequestArray = Object.keys(this.state.messages);


        let gameSelectedModal;
        if (this.state.gameSelected) {
            gameSelectedModal = this.renderGameChosenModal(this.state.gameSelected);
        } else if (this.state.showOptionsModal) {
            gameSelectedModal = this.renderOptionsModal();
        }

        let logOutBtn;
        if (typeof this.props.onLogout === 'function') {
            logOutBtn = (
                <button className="btn btn-sm float-end logout" onClick={this.props.onLogout}>Logout &#10151;</button>
            );
        }
        let optionsBtn;
        if (window.location.hash.indexOf('options=true') !== -1)
        optionsBtn = (
            <button className="btn btn-sm float-start options" onClick={this.toggleOptionsModal}>Options</button>
        );

        let mainClassName = this.state.showPlayerSelect ? 'player-select' : 'game-select';

        let subheading = this.state.showPlayerSelect ? (
            <span className="subheading-player fade-in-delay">
                Type <b>!new</b> in {this.props.channel}'s chat if you want to join the next game
            </span>
        ) : (
            <span className={`subheading-game ${(this.state.allowGameRequests === true ? 'fade-in-delay' : 'fade-out')}`} title={`Click to ${this.state.allowGameRequests === true ? 'disable' : 'enable'} game requests.`} onClick={this.toggleAllowGameRequests}>
                Type e.g. <b>"!request Blather Round"</b> in {this.props.channel}'s chat to add
            </span>
        );

        let innerContent;
        let rightColumn;

        if (this.state.showPlayerSelect) {
            innerContent = (
                <PlayerSelect
                    game={this.state.history?.[this.state.nextGameIdx]}
                    startGame={this.startGame}
                    ref={this.setPlayerSelectRef}
                />
            );
        } else {
            innerContent = gameRequestArray.map((gameName, i) =>
                <GameRequest
                    key={i}
                    gameName={gameName}
                    metadata={this.state.messages[gameName]}
                    onDelete={this.removeGame}
                    toggleLock={this.toggleLock.bind(gameName)}
                    getActivity={this.chatActivity.getStatusPromise}
                />
            );
            rightColumn = (
                <div className="right-column" width="50px">
                    <div className="wheel-wrapper fade-in">
                         <WheelComponent
                            key={this.state.counter}
                            segments={gameRequestArray}
                            segColors={this.state.colors}
                            onFinished={this.onWheelSpun}
                            isOnlyOnce={false}
                            size={250}
                            upDuration={100}
                            downDuration={1000}
                            primaryColor={"white"}
                            contrastColor={"black"}
                            fontFamily={"Arial"}
                            multilineDelimiter={' ('}
                        />
                    </div>
                </div>
            );
        }


        return (
            <div id="main-screen" className={mainClassName}>
                <nav className="main-screen-nav">
                    {optionsBtn}
                    {logOutBtn}
                </nav>
                <MessageHandler
                    addGameRequest={this.addGameRequest}
                    allowGameRequests={this.state.allowGameRequests}
                    setNextGame={this.setNextGame}
                    changeNextGameIdx={this.changeNextGameIdx}
                    startGame={this.startGame}
                    messages={this.state.messages}
                    channel={this.props.channel}
                    modList={this.props.modList}
                    access_token={this.props.access_token}
                    onMessage={this.onMessage}
                    onDelete={this.removeGame}
                    previousGames={this.state.history.slice(0, this.state.nextGameIdx)}
                    upcomingGames={this.state.history.slice(this.state.nextGameIdx)}
                    caniplayHandler={this.routePlayRequest}
                    playerExitHandler={this.routeLeaveRequest}
                    openQueueHandler={this.routeOpenQueueRequest}
                    closeQueueHandler={this.routeCloseQueueRequest}
                    clearQueueHandler={this.routeClearQueueRequest}
                    toggleAllowGameRequests={this.toggleAllowGameRequests}
                    ref={this.setMessageHandlerRef}
                />
                <div className="left-column fade-in">

                    <h1>{this.state.showPlayerSelect ? 'Seat Requests' : 'Game Requests'}</h1>
                    <h4>{subheading}</h4>

                    <div className="left-column-body">
                        <Sidebar
                            changeGameOrder={this.changeGameOrder}
                            history={this.state.history}
                            nextGameIdx={this.state.nextGameIdx}
                            changeNextGameIdx={this.changeNextGameIdx}
                            moveNextGameFwd={this.moveNextGameFwd}
                            moveNextGameBack={this.moveNextGameBack}
                            togglePlayerSelect={this.togglePlayerSelect}
                            requestMode={this.state.showPlayerSelect ? 'seat' : 'game'}
                        />
                        <div className="left-column-inner-body">
                            {innerContent}
                        </div>
                    </div>
                </div>
                {rightColumn}
                {gameSelectedModal}
            </div>
        )
    }
}
