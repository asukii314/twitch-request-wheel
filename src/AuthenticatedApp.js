import React, {Component} from 'react';
import MainScreen from './landing/MainScreen';
import LoadSpinner from './components/LoadSpinner';
import {Redirect, withRouter} from "react-router-dom";
import queryString from 'query-string'
const fetch = require('node-fetch');

class AuthenticatedApp extends Component {
    constructor() {
        super();
        this.state = {
            username: localStorage.getItem('__username'),
            user_id: localStorage.getItem('__user_id'),
            access_token: localStorage.getItem('__access_token'),
            failed_login: false
        }
        this.getAuth = this.getAuth.bind(this);
        this.getUsers = this.getUsers.bind(this);
        this.logOut = this.logOut.bind(this);
    }

    componentDidMount() {
        this._isMounted = true;
        if (!this.state.access_token) {
            return this.getAuth();
        }
        return this.getUsers(this.state.access_token).catch(e => {
            console.error(e);
            return this.getAuth(e);
        });
    }

    componentWillUnmount() {
        this._isMounted = false;
    }

    async getAuth(e) {
        if (e) {
            console.error(e);
        }
        localStorage.removeItem('__username');
        localStorage.removeItem('__user_id');
        localStorage.removeItem('__access_token');

        const queryParams = queryString.parse(this.props.location.search);
        const requestParams = new URLSearchParams({
            grant_type: 'authorization_code',
            code: queryParams.code,
            client_id: process.env.REACT_APP_TWITCH_CLIENT_ID,
            client_secret: process.env.REACT_APP_TWITCH_CLIENT_SECRET,
            redirect_uri: process.env.REACT_APP_REDIRECT_URI_NOENCODE
        });
        return await fetch(`https://id.twitch.tv/oauth2/token?${requestParams}`, {
            method: 'POST',
            headers: {
                Accept: 'application/vnd.twitchtv.v5+json'
            }
        })
        .then(r => r.json())
        .then((oauth) => {
            //console.log(oauth);  access_token, refresh_token, expires_in, scope ['...']
            if (this._isMounted) {
                if (!oauth.access_token) {
                    return this.setState({
                        failed_login: true
                    });
                }

                localStorage.setItem('__access_token', oauth.access_token);

                this.setState({
                    access_token: oauth.access_token
                });

                return this.getUsers(oauth.access_token);
            }
            return;
        })
        .catch(e => {
            console.error(e);
            if (this._isMounted) {
                return this.setState({
                    failed_login: true
                });
            }
            return;
        });
    }

    getUsers(access_token) {
        return fetch('https://api.twitch.tv/helix/users', {
            headers: {
                'Client-ID': process.env.REACT_APP_TWITCH_CLIENT_ID,
                Authorization: `Bearer ${this.state.access_token}`
            }
        })
        .then(r => r.json())
        .then(userInfo => {
            //console.log(userInfo); login [aka lowercase username?], display_name, profile_image_url, description
            localStorage.setItem('__username', userInfo.data[0].login);
            localStorage.setItem('__user_id', userInfo.data[0].id);
            return fetch(`https://api.twitch.tv/helix/moderation/moderators?broadcaster_id=${userInfo.data[0].id}`, {
                headers: {
                    'Client-ID': process.env.REACT_APP_TWITCH_CLIENT_ID,
                    Authorization: `Bearer ${this.state.access_token}`
                }
            })
            .then(r => r.json())
            .then(modInfo => {
                const modList = (!modInfo.data)
                    ? null
                    : modInfo.data.map((modObj) => (!modObj.user_name) ? null : modObj.user_name.toLowerCase()).filter(user => user);
                if (this._isMounted) {
                    return this.setState({
                        username: userInfo.data[0].login,
                        user_id: userInfo.data[0].id,
                        modList
                    });
                }
                return;
            });
        });
    }

    logOut() {
        localStorage.removeItem('__username');
        localStorage.removeItem('__user_id');
        localStorage.removeItem('__access_token');

        const requestParams = new URLSearchParams({
            client_id: process.env.REACT_APP_TWITCH_CLIENT_ID,
            token: this.state.access_token,
            redirect_uri: process.env.REACT_APP_REDIRECT_URI_NOENCODE
        });

        return fetch(`https://id.twitch.tv/oauth2/revoke?${requestParams}`, {
            method: 'POST',
            headers: {
                Accept: 'application/vnd.twitchtv.v5+json'
            }
        }).then(() => {
            return window.location.reload();
        });
    }

    render() {
        if (this.state.failed_login) {
            return (<Redirect to="/login"/>);
        }
        let mainContent = (
            <h2>
                <LoadSpinner />
                <br />
                Loading...
            </h2>
        );
        let classNames = ['authenticated-app'];
        if (this.state.username && this.state.modList) {
            mainContent = (
                <MainScreen
                    channel={this.state.username}
                    id={this.state.user_id}
                    modList={this.state.modList}
                    access_token={this.state.access_token}
                    onLogout={this.logOut}
                />
            );
        }

        return (
            <div id={classNames.join(' ')}>
                {mainContent}
            </div>
        );
    }
}

export {AuthenticatedApp};

export default withRouter(AuthenticatedApp);
