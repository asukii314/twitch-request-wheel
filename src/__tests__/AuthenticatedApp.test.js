import {AuthenticatedApp} from '../AuthenticatedApp';
import {createRenderer} from 'react-test-renderer/shallow';
import fetch from 'node-fetch';
import MainScreen from '../landing/MainScreen';
import React from 'react';
import {Redirect} from "react-router-dom";

jest.mock('../landing/MainScreen');
jest.mock('node-fetch');
jest.mock('react-router-dom', () => {
    const reactRouterDom = jest.requireActual('react-router-dom');
    return {
        ...reactRouterDom,
        Redirect: () => () => (<div>React Tooltip Mock</div>)
    };
});

describe('AuthenticatedApp', () => {
    // TODO: fix the following error from AuthenticatedApp (when importing the default class):
    // Error: Invariant failed: You should not use <withRouter(AuthenticatedApp) /> outside a <Router>
    let props = {
    	history: {
    		length: 3,
    		action: 'POP',
    		location: {
    			pathname: '/',
    			search: '?code=foobar&scope=chat:read chat:edit moderation:read whispers:edit',
    			hash: ""
    		}
    	},
    	location: {
    		pathname: '/',
    		search: '?code=foobar&scope=chat:read chat:edit moderation:read whispers:edit',
    		hash: ""
    	},
    	match: {
    		path: '/',
    		url: '/',
    		isExact: true,
    		params: {}
    	}
    };

    describe('componentDidMount', () => {
        test('should call getAuth when no access_token set in state', () => {
            let component = new AuthenticatedApp();
            jest.spyOn(component, 'getAuth').mockResolvedValue(null);

            component.componentDidMount();
            expect(component.getAuth).toHaveBeenCalledTimes(1);
        });
        test('should call getUsers when access_token found in state', () => {
            let component = new AuthenticatedApp();
            jest.spyOn(component, 'getUsers').mockResolvedValue(null);
            component.state.access_token = 'vroom-vroom-lewmon-crew';

            component.componentDidMount();

            expect(component.getUsers).toHaveBeenCalledTimes(1);
        });
        test('should call getAuth when an error is caught from getUsers', async () => {
            let component = new AuthenticatedApp();
            jest.spyOn(component, 'getAuth').mockResolvedValue(null);
            jest.spyOn(component, 'getUsers').mockRejectedValue('error stub');
            jest.spyOn(console, 'error').mockImplementation(() => {});
            component.state.access_token = 'vroom-vroom-lewmon-crew';

            await component.componentDidMount();

            expect(component.getUsers).toHaveBeenCalledTimes(1);
            expect(component.getAuth).toHaveBeenCalledTimes(1);
        });
    });

    describe('getAuth', () => {
        const props = {
            location: {
                search: 'code=54vabs9d2sd1f08pk4bjmwyjpx3iju&scope=chat%3Aread+chat%3Aedit+moderation%3Aread+whispers%3Aedit'
            }
        }
        test('should call setState and getUsers', async () => {
            jest.spyOn(window.localStorage.__proto__, 'removeItem');
            fetch.mockImplementation(() => {
                return Promise.resolve({
                    json: () => Promise.resolve({
                        access_token: 'vroom-vroom-lewmon-crew'
                    }),
                })
            });
            let component = new AuthenticatedApp(props);
            jest.spyOn(component, 'getUsers').mockResolvedValue(true);
            jest.spyOn(component, 'setState').mockImplementation(() => {});
            component._isMounted = true;
            component.props = props;

            await component.getAuth();

            expect(window.localStorage.__proto__.removeItem).toHaveBeenCalledTimes(3);
            expect(component.getUsers).toHaveBeenCalledTimes(1);
            expect(component.setState).toHaveBeenCalledWith({access_token: 'vroom-vroom-lewmon-crew'});
        });
        test('should log events if passed and only call setState if no token returned', async () => {
            jest.spyOn(console, 'error');
            jest.spyOn(window.localStorage.__proto__, 'removeItem');
            fetch.mockImplementation(() => {
                return Promise.resolve({
                    json: () => Promise.resolve({}),
                })
            });
            let component = new AuthenticatedApp(props);
            jest.spyOn(component, 'setState').mockImplementation(() => {});
            component._isMounted = true;
            component.props = props;

            await component.getAuth('error stub');

            expect(console.error).toHaveBeenCalledWith('error stub');
            expect(window.localStorage.__proto__.removeItem).toHaveBeenCalledTimes(3);
            expect(component.setState).toHaveBeenCalledWith({
                failed_login: true
            });
        });
        test('should catch any fetch errors and call setState', async () => {
            jest.spyOn(window.localStorage.__proto__, 'removeItem');
            fetch.mockImplementation(() => {
                return Promise.resolve({
                    json: () => Promise.reject({}),
                })
            });
            let component = new AuthenticatedApp(props);
            jest.spyOn(component, 'setState').mockImplementation(() => {});
            component._isMounted = true;
            component.props = props;

            await component.getAuth();

            expect(window.localStorage.__proto__.removeItem).toHaveBeenCalledTimes(3);
            expect(component.setState).toHaveBeenCalledWith({
                failed_login: true
            });
        });

        test('should not call setState if not mounted', async () => {
            jest.spyOn(window.localStorage.__proto__, 'removeItem');
            fetch.mockImplementationOnce(() => {
                return Promise.resolve({
                    json: () => Promise.resolve({}),
                })
            }).mockImplementationOnce(() => {
                return Promise.resolve({
                    json: () => Promise.reject({}),
                })
            });
            let component = new AuthenticatedApp(props);
            jest.spyOn(component, 'setState').mockImplementation(() => {});
            component.props = props;

            await component.getAuth();
            await component.getAuth();

            expect(window.localStorage.__proto__.removeItem).toHaveBeenCalledTimes(6);
            expect(component.setState).toHaveBeenCalledTimes(0);
        });
    });

    describe('getUsers', () => {
        const props = {
            location: {
                search: 'code=54vabs9d2sd1f08pk4bjmwyjpx3iju&scope=chat%3Aread+chat%3Aedit+moderation%3Aread+whispers%3Aedit'
            }
        }
        test('should call setState with broadcaster and list of mods', async () => {
            jest.spyOn(window.localStorage.__proto__, 'setItem');
            fetch.mockImplementationOnce(() => {
                return Promise.resolve({
                    json: () => Promise.resolve({
                        data: [{
                            id: '123456789',
                            login: 'sirfarewell'
                        }]
                    }),
                })
            }).mockImplementationOnce(() => {
                return Promise.resolve({
                    json: () => Promise.resolve({
                        data: [{
                            broadcaster_type: '',
                            created_at: '2019-11-18T00:47:34Z',
                            display_name: 'SirFarewell',
                            id: '123456789',
                            login: 'sirfarewell'
                        }, {
                            user_name: 'HerooftheSprites'
                        }],
                        pagination: {}
                    }),
                })
            });


            let component = new AuthenticatedApp(props);
            jest.spyOn(component, 'setState').mockImplementation(() => {});
            component._isMounted = true;
            component.props = props;

            await component.getUsers();

            expect(window.localStorage.__proto__.setItem).toHaveBeenCalledTimes(2);
            expect(component.setState).toHaveBeenCalledWith({
                username: 'sirfarewell',
                user_id: '123456789',
                modList: ['heroofthesprites']
            });
        });
        test('should not call setState if not mounted', async () => {
            jest.spyOn(window.localStorage.__proto__, 'setItem');
            fetch.mockImplementationOnce(() => {
                return Promise.resolve({
                    json: () => Promise.resolve({
                        data: [{
                            id: '123456789',
                            login: 'sirfarewell'
                        }]
                    }),
                })
            }).mockImplementationOnce(() => {
                return Promise.resolve({
                    json: () => Promise.resolve({}),
                })
            });


            let component = new AuthenticatedApp(props);
            jest.spyOn(component, 'setState').mockImplementation(() => {});
            component.props = props;

            await component.getUsers();

            expect(window.localStorage.__proto__.setItem).toHaveBeenCalledTimes(2);
            expect(component.setState).toHaveBeenCalledTimes(0);
        });
    });

    describe('logOut', () => {
        const {location} = window;

        beforeAll(() => {
            delete window.location;
            window.location = {reload: jest.fn()};
        });

        afterAll(() => {
            window.location = location;
        });
        test('should log out of api and reload window', async () => {
            jest.spyOn(window.localStorage.__proto__, 'removeItem');
            fetch.mockImplementation(() => {
                return Promise.resolve({
                    json: () => Promise.resolve({}),
                })
            });
            let component = new AuthenticatedApp();
            await component.logOut();

            expect(window.localStorage.__proto__.removeItem).toHaveBeenCalledTimes(3);
            expect(window.location.reload).toHaveBeenCalled();
        });
    });

    describe('render', () => {
        test('should render with MainScreen', () => {
            const shallowRenderer = createRenderer();
            shallowRenderer.render(<AuthenticatedApp {...props} />);
            let instance = shallowRenderer.getMountedInstance();
            instance.setState({
                access_token: 'yadayadayada',
                failed_login: false,
                modList: [],
                username: 'sirgoosewell'
            });
            let component = shallowRenderer.getRenderOutput();
            expect(component.props.children.type).toBe(MainScreen);
            expect(component).toMatchSnapshot();
            shallowRenderer.unmount();
        });
        test('should render with Redirect on failed login', () => {
            const shallowRenderer = createRenderer();
            shallowRenderer.render(<AuthenticatedApp {...props} />);
            let instance = shallowRenderer.getMountedInstance();
            instance.setState({
                access_token: null,
                failed_login: true,
                username: null
            });
            let component = shallowRenderer.getRenderOutput();
            expect(component.type).toBe(Redirect);
            expect(component).toMatchSnapshot();
            shallowRenderer.unmount();
        });
        test('should render with null', () => {
            const shallowRenderer = createRenderer();
            shallowRenderer.render(<AuthenticatedApp {...props} />);
            let instance = shallowRenderer.getMountedInstance();
            instance.setState({
                access_token: null,
                failed_login: false,
                username: null
            });
            let component = shallowRenderer.getRenderOutput();
            expect(component.props.children).toBeDefined();
            expect(component).toMatchSnapshot();
            shallowRenderer.unmount();
        });
    });
});
