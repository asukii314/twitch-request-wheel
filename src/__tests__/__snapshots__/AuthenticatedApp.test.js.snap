// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthenticatedApp render should render with MainScreen 1`] = `
<div
  id="authenticated-app"
>
  <MainScreen
    access_token="yadayadayada"
    channel="sirgoosewell"
    modList={Array []}
    onLogout={[Function]}
  />
</div>
`;

exports[`AuthenticatedApp render should render with Redirect on failed login 1`] = `
<Redirect
  to="/login"
/>
`;

exports[`AuthenticatedApp render should render with null 1`] = `
<div
  id="authenticated-app"
>
  <h2>
    <LoadSpinner />
    <br />
    Loading...
  </h2>
</div>
`;
